pkg/datapath/loader/tc.go

```diff
@@ -4,19 +4,41 @@
 package loader
 
 import (
+	"errors"
 	"fmt"
 	"strings"
 
 	"github.com/vishvananda/netlink"
 	"golang.org/x/sys/unix"
 
 	"github.com/cilium/ebpf"
+	"github.com/cilium/ebpf/link"
 
 	"github.com/cilium/cilium/pkg/option"
 )
 
-// attachSKBProgram attaches prog to device using legacy tc.
-func attachSKBProgram(device netlink.Link, prog *ebpf.Program, progName, bpffsDir string, parent uint32) error {
+// attachSKBProgram attaches prog to device using tcx if available and enabled,
+// or legacy tc as a fallback.
+func attachSKBProgram(device netlink.Link, prog *ebpf.Program, progName, bpffsDir string, parent uint32, tcxEnabled bool) error {
+	if tcxEnabled {
+		// Attach using tcx if available. This is seamless on interfaces with
+		// existing tc programs since attaching tcx disables legacy tc evaluation.
+		err := upsertTCXProgram(device, prog, progName, bpffsDir, parent)
+		if err == nil {
+			// Created tcx link, clean up any leftover legacy tc attachments.
+			if err := removeTCFilters(device, parent); err != nil {
+				return fmt.Errorf("legacy tc cleanup after attaching tcx program %s: %w", progName, err)
```

Comment:
if we throw an error here will upsertTCXProgram leave state changed? Should we revert change to state when returning an error if this is the case?

Reply:
(I will follow-up on this.)

Reply:
Yeah, let's not bail out if tcx already attached successfully and removing the legacy filter failed somehow.